src/coreclr/jit/assertionprop.cpp

```diff
@@ -2238,18 +2260,24 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
             if (con >= 0)
             {
                 AssertionDsc dsc;
-
-                // For arr.Length != 0, we know that 0 is a valid index
-                // For arr.Length == con, we know that con - 1 is the greatest valid index
+                bool         useNextEdge;
                 if (con == 0)
                 {
+                    // arr.Length != 0 -> Then: idx = 0        Else: no assertion
+                    // arr.Length == 0 -> Then: no assertion,  Else: idx = 0
+                    //
```

Comment:
```suggestion
                    // arr.Length != 0 -> Then: arr.Length == 0        Else: no assertion
                    // arr.Length == 0 -> Then: no assertion,  Else: arr.Length == 0
                    //
```
? Or have I misunderstood what kind of assertion this is creating?

Reply:
> ? Or have I misunderstood what kind of assertion this is creating?

so `if (arr.Length != 0)` sort of simulates what the following assertion (`OAK_NO_THROW`) does:
```
_ = arr[0]
```
so it basically says "Idx=0" already accessed this array, hence, it's Idx in my comment.

We have 2 options here:
1) Redesign `O1K_LCLVAR` to work on arbitrary VNs (in global assert prop) rather than requiring LCLVAR only.
2) For "arr.Length != 0" we can create whatever assertion we generate for "arr.Length > 0" (I tried that, but it involves creating artificial VNs)

Reply:
> so it basically says "Idx=0" already accessed this array,

Shouldn't this be creating an `OAK_NO_THROW` assertion then? AFAIK that's the kind of assertion we create on an actual array access.

Reply:
> > so it basically says "Idx=0" already accessed this array,
> 
> Shouldn't this be creating an `OAK_NO_THROW` assertion then? AFAIK that's the kind of assertion we create on an actual array access.

Yes. let me see if I can change it..

Reply:
@jakobbotsch so I've changed it to `OAK_NO_THROW ` via
```cpp
AssertionDsc dsc = {};
bool         useNextEdge;
if (con == 0)
{
    // arr.Length != 0 -> Create arr[0] assertion in the "then" block
    // arr.Length == 0 -> Create arr[0] assertion in the "else" block
    //
    dsc.op1.bnd.vnIdx = vnStore->VNForIntCon(0);
    useNextEdge       = relop->OperIs(GT_EQ);
}
else
{
    // arr.Length == 10 -> Create arr[9] assertion in the "then" block
    // arr.Length != 10 -> Create arr[9] assertion in the "else" block
    //
    dsc.op1.bnd.vnIdx = vnStore->VNForIntCon(con - 1);
    useNextEdge       = relop->OperIs(GT_NE);
}

dsc.assertionKind = OAK_NO_THROW;
dsc.op1.vn        = optConservativeNormalVN(relop);
dsc.op1.kind      = O1K_ARR_BND;
dsc.op1.bnd.vnLen = op1VN;
dsc.op2.kind      = O2K_INVALID;
dsc.op2.vn        = ValueNumStore::NoVN;

// when con is not zero, create an assertion on the arr.Length == con edge
// when con is zero, create an assertion on the arr.Length != 0 edge
AssertionIndex index = optAddAssertion(&dsc);
return useNextEdge ? AssertionInfo::ForNextEdge(index) : index;
```

it worked, but produced more regressions than improvements. Looks like it's caused by rangecheck's merge assertion logic. Do you mind if I keep it as is then? My PR doesn't change this logic, just adds a comment. I'll look into rangecheck's merge assertion logic separately (already found an issue)

Reply:
I don't mind, but the comments should be fixed: we should point out that this is only used for range check, that the `vnIdx` and `vnBnd` parts are unused (set them to `NoVN`), and the comments should refer to the fact that the assertion created is really `arr.Length == 0` etc.

Reply:
> the `vnIdx` parts are unused (set it to `NoVN`), and the comments should refer to the fact that the assertion created is really `arr.Length == 0` etc.


sorry but I don't understand how it is supposed to work for e.g. `if (arr.Length == 10)`
if I set `vnIdx` to NoVN where the 10 (or 9) constant goes? given that op2 is 0.

Reply:
The `10` should be in op2. That's the bug. `OAK_EQUAL` is saying that `op1 == op2`, so it should be creating an assertion that reflects that `op1 == 10` here. We also don't have a way to encode arbitrary VN for the LHS, which we should add, instead of misusing `O1K_ARR_BND`.

Range check uses the VN to extract the constant, so it does not hit the bug:
https://github.com/dotnet/runtime/blob/0170d78578a04206542f32e4db3eaa88a5384973/src/coreclr/jit/rangecheck.cpp#L747-L754

Reply:
Feel free to just update the comments (they shouldn't refer to `idx` since it's meaningless here) and I will approve

Reply:
Closing the thread as this change is no more